lib/utils/fs.go

```diff
@@ -84,27 +85,59 @@ func IsDir(path string) bool {
 
 // NormalizePath normalises path, evaluating symlinks and converting local
 // paths to absolute
-func NormalizePath(path string) (string, error) {
+func NormalizePath(path string, evaluateSymlinks bool) (string, error) {
 	s, err := filepath.Abs(path)
 	if err != nil {
 		return "", trace.ConvertSystemError(err)
 	}
-	abs, err := filepath.EvalSymlinks(s)
-	if err != nil {
-		return "", trace.ConvertSystemError(err)
+	if evaluateSymlinks {
+		s, err = filepath.EvalSymlinks(s)
+		if err != nil {
+			return "", trace.ConvertSystemError(err)
+		}
 	}
-	return abs, nil
+	return s, nil
+}
+
+// OpenFileAllowingSymlinks will open a file, if it's a symlink the file referenced by the link will be associated to
+// the returned os.File.  This will return an error if the file is not found or is a directory.
+func OpenFileAllowingSymlinks(path string) (*os.File, error) {
+	return openFile(path, true)
+}
+
+// OpenFileNoSymlinks will open a file, ensuring it's an actual file and not a directory or symlink.
+func OpenFileNoSymlinks(path string) (*os.File, error) {
+	return openFile(path, false)
 }
 
-// OpenFile opens  file and returns file handle
-func OpenFile(path string) (*os.File, error) {
-	newPath, err := NormalizePath(path)
+// openFile opens a file and returns file handle. In general symlinks should not be allowed to reduce the risk of
+// privilege escalation from Teleports elevated privileges to potentially less privileged users accidentally.
+func openFile(path string, allowSymlink bool) (*os.File, error) {
+	newPath, err := NormalizePath(path, allowSymlink)
 	if err != nil {
 		return nil, trace.Wrap(err)
 	}
-	fi, err := os.Stat(newPath)
-	if err != nil {
-		return nil, trace.ConvertSystemError(err)
+	var fi os.FileInfo
+	if allowSymlink {
+		fi, err = os.Stat(newPath)
+		if err != nil {
+			return nil, trace.ConvertSystemError(err)
+		}
+	} else {
+		isAbs := filepath.IsAbs(newPath)
+		components := strings.Split(newPath, string(os.PathSeparator))
+		for i := 1; i <= len(components); i++ {
+			subPath := filepath.Join(components[:i]...)
+			if isAbs {
+				subPath = string(os.PathSeparator) + subPath
+			}
```

Comment:
Suggestion: Use a for-range, Join addictively, check first path component, spacing.

```suggestion
		components := strings.Split(newPath, string(os.PathSeparator))
		var subpath string
		for _, p := range components {
			subpath = filepath.Join(subpath, p)

			// The first component in an absolute path is empty.
			if subpath == "" {
				subPath = string(os.PathSeparator)
			}

```

Reply:
It's unclear to me why this is better, let me know your thoughts

Reply:
for-ranges are usually preferable as there is no need to deal with `i`. This ties in with a previous comment.

The snippet also replaces filepath.IsAbs with an empty string check, which I'd expect to be much cheaper - not a big deal here, but why not do it?

Other than those, it's a toss, hence the "suggestion" tag.

Reply:
Updated, you convinced me this is more readable.  Thank you!

Reply:
Thanks!